### PR TITLE
obs-ffmpeg: Remove stream key from ffmpeg log 

### PR DESCRIPTION
**Pull request by mv2731**
My prior code for streaming HLS from OBS had one major security flaw:
the stream key is printed to the output log both by OBS and by the underlying
ffmpeg log. Remove channel stream key and replace with "{stream_key}" for
both OBS and the ffmpeg log. To do so for the ffmpeg log, write a log function
to be called as a callback and check if stream key is present in the line. If so, 
censor it. Also, add colors to the warning statements of the ffmpeg log to 
resemble the original as closely as possible.

### CODE REVIEW

File: plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c [hunk 1]
```diff
@@ -471,7 +527,7 @@ static inline int open_output_file(struct ffmpeg_mux *ffm)
 		ret = avio_open(&ffm->output->pb, ffm->params.file,
 				AVIO_FLAG_WRITE);
 		if (ret < 0) {
-			fprintf(stderr, "Couldn't open '%s', %s",
+			fprintf(stderr, "Couldn't open '%s', %s\n",
 				ffm->params.file, av_err2str(ret));
```
Comment (ushadow) [inline plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c]: Does this file contain stream key?

File: plugins/obs-ffmpeg/obs-ffmpeg-mux.c [hunk 2]
```diff
@@ -436,6 +456,11 @@ static int deactivate(struct ffmpeg_muxer *stream, int code)
 		os_atomic_set_bool(&stream->active, false);
 		os_atomic_set_bool(&stream->sent_headers, false);
 
+		service = obs_output_get_service(stream->output);
+		if (!service)
+			return false;
+		stream_key = obs_service_get_key(service);
+		dstr_replace(&stream->path, stream_key, "{stream_key}");
```
Comment (ushadow) [inline plugins/obs-ffmpeg/obs-ffmpeg-mux.c]: Don't directly change stream->path. Create a temporary variable.

File: plugins/obs-ffmpeg/obs-ffmpeg-mux.c [hunk 3]
```diff
@@ -217,6 +217,23 @@ static void log_muxer_params(struct ffmpeg_muxer *stream, const char *settings)
...
+	obs_data_t *service;
+	const char *path_str;
+	const char *stream_key;
+
+	service = obs_output_get_service(stream->output);
+	if (!service) {
+		dstr_catf(cmd, "\"%s\" ", "");
+		return;
+	}
+	path_str = obs_service_get_url(service);
```
Comment (ushadow) [inline plugins/obs-ffmpeg/obs-ffmpeg-mux.c]: path_str is not needed.

File: plugins/obs-ffmpeg/obs-ffmpeg-mux.c [hunk 4]
```diff
@@ -217,6 +217,23 @@ static void log_muxer_params(struct ffmpeg_muxer *stream, const char *settings)
 	av_dict_free(&dict);
 }
 
+static void add_stream_key(struct dstr *cmd, struct ffmpeg_muxer *stream)
+{
+	obs_data_t *service;
+	const char *path_str;
+	const char *stream_key;
+
+	service = obs_output_get_service(stream->output);
```
Comment (ushadow) [inline plugins/obs-ffmpeg/obs-ffmpeg-mux.c:226]: Create a function for getting the stream key so we don't repeat the same code.

File: plugins/obs-ffmpeg/obs-ffmpeg-mux.c [hunk 5]
```diff
@@ -436,6 +456,11 @@ static int deactivate(struct ffmpeg_muxer *stream, int code)
 		os_atomic_set_bool(&stream->active, false);
 		os_atomic_set_bool(&stream->sent_headers, false);
 
+		service = obs_output_get_service(stream->output);
+		if (!service)
```
Comment (ushadow) [inline plugins/obs-ffmpeg/obs-ffmpeg-mux.c]: You can't just return here. The rest of the function still needs to be executed.

File: plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c [hunk 6]
```diff
@@ -200,6 +210,48 @@ static bool get_audio_params(struct audio_params *audio, int *argc,
...
+
+	struct dstr log_message = {0};
+	stream_key = global_stream_key;
+
+	vsnprintf(out, sizeof(out), format, args);
+	dstr_copy(&log_message, out);
+
+	if ((strcmp(stream_key, "") != 0) &&
+	    (dstr_find(&log_message, stream_key)))
+		dstr_replace(&log_message, stream_key, "{stream_key}");
```
Comment (ushadow) [inline plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c]: Do we need dstr_find? If stream_key is not present, does dstr_replace return the original string?

File: plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c [hunk 7]
```diff
@@ -200,6 +210,48 @@ static bool get_audio_params(struct audio_params *audio, int *argc,
...
+
+	strcpy(out, log_message.array);
```
Comment (ushadow) [inline plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c]: Is this necessary?

File: plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c [hunk 8]
```diff
@@ -27,8 +27,16 @@
 #include <stdlib.h>
 #include "ffmpeg-mux.h"
 
+#include <obs-module.h>
```
Comment (ushadow) [inline plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c]: Is obs-module.h needed?

Comment (mv2731) [reply plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c]: You're right it's not! I think I needed it in an earlier version of the code but I no longer need it.

File: plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c (hunk 6 above)
Comment (mv2731) [reply plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c]: Reread dstr_find and I think we don't need it. I was trying to be overly cautious.

File: plugins/obs-ffmpeg/obs-ffmpeg-mux.c (hunk 2 above)
Comment (mv2731) [reply plugins/obs-ffmpeg/obs-ffmpeg-mux.c]: Done.

File: plugins/obs-ffmpeg/obs-ffmpeg-mux.c (hunk 5 above)
Comment (mv2731) [reply plugins/obs-ffmpeg/obs-ffmpeg-mux.c]: Good point. Fixed.

File: plugins/obs-ffmpeg/obs-ffmpeg-mux.c (hunk 4 above)
Comment (mv2731) [reply plugins/obs-ffmpeg/obs-ffmpeg-mux.c:226]: Done.

File: plugins/obs-ffmpeg/obs-ffmpeg-mux.c (hunk 2 above)
Comment (mv2731) [reply plugins/obs-ffmpeg/obs-ffmpeg-mux.c]: Done.

**Pull request closed without merging**